🤖 AGENTS.md consolidation and updates (#7194)
Comfy-Org/ComfyUI_frontend: AGENTS.md

# Repository Guidelines

## Project Structure & Module Organization

- Source: `src/`
  - Vue 3.5+
  - TypeScript
  - Tailwind 4
  - Key areas:
    - `components/`
    - `views/`
    - `stores/` (Pinia)
    - `composables/`
    - `services/`
    - `utils/`
    - `assets/`
    - `locales/`
- Routing: `src/router.ts`,
- i18n: `src/i18n.ts`,
- Entry Point: `src/main.ts`.
- Tests:
  - unit/component in `tests-ui/` and `src/**/*.test.ts`
  - E2E (Playwright) in `browser_tests/**/*.spec.ts`
- Public assets: `public/`
- Build output: `dist/`
- Configs
  - `vite.config.mts`
  - `vitest.config.ts`
  - `playwright.config.ts`
  - `eslint.config.ts`
  - `.prettierrc`
  - etc.

## Monorepo Architecture

> The project now uses **Nx** for build orchestration and task management:

Reviewer: ```suggestion
The project uses **Nx** for build orchestration and task management:
```

> - **Task Orchestration**: Commands like `dev`, `build`, `lint`, and `test:browser` run via Nx

Reviewer: Wonder if we should explain how to do browser tests locally like

1. Stash, generate baseline for the test
2. Stash pop, run the test

- `pnpm test:unit`: Run Vitest unit tests
- `pnpm test:browser`: Run Playwright E2E tests (`browser_tests/`)
- `pnpm lint` / `pnpm lint:fix`: Lint (ESLint)
- `pnpm format` / `pnpm format:check`: Prettier
- `pnpm typecheck`: Vue TSC type checking

## Coding Style & Naming Conventions

- Language:
  - TypeScript (exclusive, no new JavaScript)
  - Vue 3 SFCs (`.vue`)
    - Composition API only
  - Tailwind 4 styling
    - Avoid `<style>` blocks
- Style: (see `.prettierrc`)
  - Indent 2 spaces
  - single quotes
  - no trailing semicolons
  - width 80
- Imports:
  - sorted/grouped by plugin
  - run `pnpm format` before committing
- ESLint:
  - Vue + TS rules
  - no floating promises
  - unused imports disallowed
  - i18n raw text restrictions in templates
- Naming:
  - Vue components in PascalCase (e.g., `MenuHamburger.vue`)
  - composables `useXyz.ts`
  - Pinia stores `*Store.ts`

## Testing Guidelines

- Frameworks:
  - Vitest (unit/component, happy-dom)
  - Playwright (E2E)
- Test files:
  - Unit/Component: `**/*.test.ts`
  - E2E: `browser_tests/**/*.spec.ts`
  - Litegraph Specific: `src/lib/litegraph/test/`
- Coverage: text/json/html reporters enabled

>   - aim to cover critical logic and new features

Author: YES. YES I DO.

- Playwright:
  - optional tags like `@mobile`, `@2x` are respected by config

## Commit & Pull Request Guidelines

- Commits: Use `[skip ci]` for locale-only updates when appropriate
- PRs: Include clear description, linked issues (`- Fixes #123`), and screenshots/GIFs for UI changes

> - Quality gates: `pnpm lint`, `pnpm typecheck`, and relevant tests must pass

Reviewer: ```suggestion
- Quality gates: `pnpm lint`, `pnpm typecheck`, `pnpm knip` and relevant tests must pass
- Never use `--no-verify` to bypass failing tests. Identify the issue and present root cause analysis and possible solutions if you are unable to solve quickly yourself.
```

> - Keep PRs focused and small

Reviewer: ```suggestion
- Keep PRs focused and small. If it looks like the current changes will have 300+ lines of non-test code, suggest ways it could be broken into multiple PRs
```

  - Example:

    ```typescript
    const { nodes, showTotal = true } = defineProps<{
    nodes: ApiNodeCost[]
    showTotal?: boolean
    }>()
    ```

  - Do not use `withDefaults` or runtime props declaration
  - Do not import Vue macros unnecessarily
  - Prefer `useModel` to separately defining a prop and emit
  - Be judicious with addition of new refs or other state
    - If it's possible to accomplish the design goals with just a prop, don't add a `ref`
    - If it's possible to use the `ref` or prop directly, don't add a `computed`
    - If it's possible to use a `computed` to name and reuse a derived value, don't use a `watch`

## Development Guidelines

1. Leverage VueUse functions for performance-enhancing styles
2. Use es-toolkit for utility functions
3. Use TypeScript for type safety
4. Implement proper props and emits definitions
5. Utilize Vue 3's Teleport component when needed
6. Use Suspense for async components
7. Implement proper error handling
8. Follow Vue 3 style guide and naming conventions
9. Use Vite for fast development and building
10. Use vue-i18n in composition API for any string literals. Place new translation entries in src/locales/en/main.json

> 11. Avoid new usage of PrimeVue components

Reviewer: ```suggestion
11. Avoid new usage of PrimeVue components
12. Write tests for all changes, especially bug fixes to catch future regressions
13. Write code that is expressive and self-documenting to the furthest degree possible. This reduces the need for code comments which can get out of sync with the code itself. Try to avoid comments unless absolutely necessary
14. Whenever a new piece of code is written, the author should ask themselves 'is there a simpler way to introduce the same functionality?'. If the answer is yes, the simpler course should be chosen.
15. Refactoring should be used to make complex code simpler.
```

- Prefer running single tests for performance
- Use --help for unfamiliar CLI tools

## GitHub Integration

When referencing Comfy-Org repos:

1. Check for local copy
2. Use GitHub API for branches/PRs/metadata
3. Curl GitHub website if needed

## Settings and Feature Flags Quick Reference

### Settings Usage

```typescript
const settingStore = useSettingStore()
const value = settingStore.get('Comfy.SomeSetting')     // Get setting
await settingStore.set('Comfy.SomeSetting', newValue)   // Update setting
```

### Dynamic Defaults

```typescript
{
  id: 'Comfy.Example.Setting',
  defaultValue: () => window.innerWidth < 1024 ? 'small' : 'large'  // Runtime context
}
```

### Version-Based Defaults

```typescript
{
  id: 'Comfy.Example.Feature',
  defaultValue: 'legacy',
  defaultsByInstallVersion: { '1.25.0': 'enhanced' }  // Gradual rollout
}
```

### Feature Flags

```typescript
if (api.serverSupportsFeature('feature_name')) {  // Check capability
  // Use enhanced feature
}
const value = api.getServerFeature('config_name', defaultValue)  // Get config
```

**Documentation:**

- Settings system: `docs/SETTINGS.md`

> - Feature flags system: `docs/FEATURE_FLAGS.md`

Reviewer: Agents can find this easily

## Common Pitfalls

- NEVER use `any` type - use proper TypeScript types
- NEVER use `as any` type assertions - fix the underlying type issue
- NEVER use `--no-verify` flag when committing
- NEVER delete or disable tests to make them pass
- NEVER circumvent quality checks

> - NEVER use `dark:` or `dark-theme:` tailwind variants. Instead use a semantic value from the `style.css` theme, e.g. `bg-node-component-surface`

Reviewer: ```suggestion
- NEVER use the `dark:` tailwind variant. Instead use a semantic value from the `style.css` theme, e.g. `bg-node-component-surface`
```
